### PR TITLE
Add stack based queue

### PR DESCRIPTION
**Pull request by InKahootz**
Added a stack-based queue implementation

- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have added comments to hard-to-understand areas of my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have made corresponding changes to the comments
- [x] I have made corresponding changes to the README.md

### CODE REVIEW

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures.Tests/StackBasedQueueTests.cs  2
- DataStructures/StackBasedQueue.cs  5
         

Clones added
============
- DataStructures.Tests/StackBasedQueueTests.cs  8
- DataStructures/StackBasedQueue.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6288807&bid=20460320)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures.Tests/StackBasedQueueTests.cs  2
- DataStructures/StackBasedQueue.cs  5
         

Clones added
============
- DataStructures.Tests/StackBasedQueueTests.cs  8
- DataStructures/StackBasedQueue.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6288807&bid=20460320)

File: DataStructures/StackBasedQueue.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                }
+            }
+
+            return output.Peek();
+        }
+
+        /// <summary>
+        /// Adds an item at the last position in the queue.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">Thrown if the queue is full.</exception>
```
Comment (siriak) [inline DataStructures/StackBasedQueue.cs:92]: Is it really thrown here?

Comment (InKahootz) [reply DataStructures/StackBasedQueue.cs]: I was unsure if I should keep if there or not since it's the same in the ListBasedQueue.cs.

Fixed.

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures.Tests/StackBasedQueueTests.cs  2
- DataStructures/StackBasedQueue.cs  5
         

Clones added
============
- DataStructures.Tests/StackBasedQueueTests.cs  8
- DataStructures/StackBasedQueue.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6288807&bid=20460320)

File: DataStructures/StackBasedQueue.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace DataStructures
+{
+    /// <summary>
+    /// Implementation of a stack based queue. FIFO style.
```
Comment (siriak) [inline DataStructures/StackBasedQueue.cs:7]: Could you add a note that Enqueue is O(1) and Dequeue is amortized O(1)?

**Review by siriak — DISMISSED** [review]
Thanks, looks good! One little comment and you're good to go!

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures.Tests/StackBasedQueueTests.cs  2
- DataStructures/StackBasedQueue.cs  5
         

Clones added
============
- DataStructures.Tests/StackBasedQueueTests.cs  8
- DataStructures/StackBasedQueue.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6288807&bid=20460320)

**Review by siriak — APPROVED** [review]
Thanks!

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures.Tests/StackBasedQueueTests.cs  2
- DataStructures/StackBasedQueue.cs  5
         

Clones added
============
- DataStructures.Tests/StackBasedQueueTests.cs  8
- DataStructures/StackBasedQueue.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6288807&bid=20460320)

**Pull request merged**